include/dynamatic/Transforms/Speculation/SpeculationPlacement.h

```diff
@@ -0,0 +1,108 @@
+//===- SpeculationPlacement.h - Speculation units placement -----*- C++ -*-===//
+//
+// Dynamatic is under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares the data structures needed for speculation algorithms
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef DYNAMATIC_TRANSFORMS_SPECULATION_PLACEMENT_H
+#define DYNAMATIC_TRANSFORMS_SPECULATION_PLACEMENT_H
+
+#include "circt/Dialect/Handshake/HandshakeOps.h"
+#include "dynamatic/Support/DynamaticPass.h"
+#include "dynamatic/Support/LLVM.h"
+#include "dynamatic/Support/Logging.h"
+#include "dynamatic/Support/TimingModels.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/Hashing.h"
+#include <map>
+#include <string>
+#include <unordered_set>
+#include <vector>
+
+namespace dynamatic {
+namespace speculation {
+
+struct PlacementNames {
+  StringRef srcOpName;
+  StringRef dstOpName;
+};
+
+struct OpPlacement {
+  Value srcOpResult;
+  Operation *dstOp;
+
+  bool operator==(const OpPlacement &other) const;
+
+  struct Hash {
+    std::size_t operator()(const OpPlacement &p) const;
+  };
+};
+
+typedef std::unordered_set<OpPlacement, OpPlacement::Hash> PlacementList;
+
+class SpeculationPlacements {
+private:
+  OpPlacement speculator;
+  PlacementList saves;
```

Comment:
These do not seem to be used as sets at all (e.g., no `contains` query as far as I see). If not, a vector is simpler and better (`SmallVector<OpOperand&>` with the `OpPlacement` change described above).

Reply:
The sets are being in code that is yet to be pushed, in methods `containsSave`, `containsCommit`, `eraseSave` and `eraseCommit`. A vector would make the implementation of these methods a bit longer. I suggest to keep using the set, what do you think?

Reply:
Sounds good, thanks for clarifying. You can keep them as is then.